File: backend/agent-service/app/schemas/analysis.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator


FitLevel = Literal["high", "medium", "low"]
MatchLevel = Literal["strong", "partial", "missing"]
MatchType = Literal["exact", "alias", "taxonomy", "semantic", "none"]
SkillPriority = Literal["required", "preferred", "unknown"]
EvidenceSource = Literal["experience", "project", "skills", "mention"]


class LanguageInfo(BaseModel):
    primary_language: str = "unknown"
    detected_languages: list[str] = Field(default_factory=list)
    is_mixed_language: bool = False
    confidence: float = Field(default=0.0, ge=0, le=1)


class LocalizedEvidence(BaseModel):
    original_text: str = Field(..., exclude=True)
    source_language: str
    english_text: str | None = None


class AgentAnalyzeRequest(BaseModel):
    cv_text: str = Field(..., min_length=50)
    jd_text: str = Field(..., min_length=50)

    @field_validator("cv_text", "jd_text")
    @classmethod
    def normalize_text(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if len(normalized) < 50:
            raise ValueError("Text must contain at least 50 non-whitespace characters.")
        return normalized


class CVEvidence(BaseModel):
    source: EvidenceSource
    text: str
    strength: float = Field(..., ge=0, le=1)
    source_language: str = "en"
    original_text: str | None = Field(default=None, exclude=True)


class JDSkillRequirement(BaseModel):
    skill: str
    classification: SkillPriority = "unknown"
    critical: bool = False
    evidence: str | None = None
    original_evidence: str | None = Field(default=None, exclude=True)
    english_evidence: str | None = None
    base_weight: float = Field(..., gt=0, le=1)


class SkillMatch(BaseModel):
    jd_skill: str
    resume_skill: str | None = None
    similarity: float = Field(..., ge=0, le=1)
    match_level: MatchLevel
    importance: float = Field(default=1.0, gt=0)
    classification: SkillPriority = "required"
    match_type: MatchType = "none"
    evidence: list[CVEvidence] = Field(default_factory=list)
    evidence_score: float = Field(default=0.0, ge=0, le=1)


class ComponentWeight(BaseModel):
    component: str
    configured_weight: float = Field(..., ge=0, le=1)
    enabled: bool
    effective_weight: float = Field(..., ge=0, le=1)


class ScoreBreakdown(BaseModel):
    required_skill_score: float = Field(..., ge=0, le=100)
    preferred_skill_score: float = Field(..., ge=0, le=100)
    experience_relevance_score: float = Field(..., ge=0, le=100)
    project_relevance_score: float = Field(..., ge=0, le=100)
    education_cert_relevance_score: float = Field(..., ge=0, le=100)
    unknown_skill_score: float = Field(default=0.0, ge=0, le=100)
    skill_coverage_score: float = Field(default=0.0, ge=0, le=100)
    component_weights: list[ComponentWeight] = Field(default_factory=list)


class ParsedCV(BaseModel):
    language_info: LanguageInfo = Field(default_factory=LanguageInfo)
    skills: list[str] = Field(default_factory=list)
    experience_summary: str | None = None
    experience_evidence: list[str] = Field(default_factory=list)
    projects: list[str] = Field(default_factory=list)
    education: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    localized_evidence: list[LocalizedEvidence] = Field(default_factory=list)


class ParsedJD(BaseModel):
    language_info: LanguageInfo = Field(default_factory=LanguageInfo)
    required_skills: list[str] = Field(default_factory=list)
    preferred_skills: list[str] = Field(default_factory=list)
    unknown_skills: list[str] = Field(default_factory=list)
    skill_requirements: list[JDSkillRequirement] = Field(default_factory=list)
    job_title: str | None = None
    responsibilities: list[str] = Field(default_factory=list)
    experience_requirements: list[str] = Field(default_factory=list)
    project_requirements: list[str] = Field(default_factory=list)
    education_requirements: list[str] = Field(default_factory=list)
    domain_keywords: list[str] = Field(default_factory=list)


class SkillWeightSource(BaseModel):
    skill: str
    jd_classification: SkillPriority
    jd_base_weight: float = Field(..., gt=0, le=1)
    final_weight: float = Field(..., gt=0, le=1)
    cv_evidence_score: float = Field(..., ge=0, le=1)
    contribution: float = Field(..., ge=0, le=1)


class ScoreAdjustment(BaseModel):
    rule: str
    reason: str
    score_before: float = Field(..., ge=0, le=100)
    score_after: float = Field(..., ge=0, le=100)


class AgentAnalyzeResponse(BaseModel):
    fit_score: float = Field(..., ge=0, le=100)
    fit_level: FitLevel
    score_breakdown: ScoreBreakdown

    parsed_cv: ParsedCV
    parsed_jd: ParsedJD

    skill_matches: list[SkillMatch] = Field(default_factory=list)
    matched_skills: list[str] = Field(default_factory=list)
    missing_skills: list[str] = Field(default_factory=list)

    cv_improvement_suggestions: list[str] = Field(default_factory=list)
    cover_letter: str | None = None
    learning_roadmap: list[str] | None = None

    raw_fit_score: float = Field(default=0.0, ge=0, le=100)
    score_confidence: float = Field(default=0.0, ge=0, le=1)
    scoring_version: str = "phase6-v2"
    skill_weight_sources: list[SkillWeightSource] = Field(default_factory=list)
    score_adjustments: list[ScoreAdjustment] = Field(default_factory=list)
    analysis_warnings: list[str] = Field(default_factory=list)

File: frontend/web/src/types/analysis.ts
export type SessionStatus = "processing" | "completed" | "failed";

export type FitLevel = "high" | "medium" | "low";

export type MatchLevel = "strong" | "partial" | "missing";
export type SkillPriority = "required" | "preferred" | "unknown";
export type MatchType = "exact" | "alias" | "taxonomy" | "semantic" | "none";

export interface AnalyzeStartResponse {
  session_id: string;
  status: "processing";
}

export interface SessionError {
  code: string;
  message: string;
  detail?: unknown;
}

export interface DocumentParseResult {
  filename: string;
  document_type: string | null;
  content_type: string | null;
  file_size_bytes: number;
  page_count: number;
  text_length: number;
  warnings: string[];
}

export interface ScoreBreakdown {
  required_skill_score: number;
  preferred_skill_score: number;
  experience_relevance_score: number;
  project_relevance_score: number;
  education_cert_relevance_score: number;
  unknown_skill_score: number;
  skill_coverage_score: number;
  component_weights: ComponentWeight[];
}

export interface ComponentWeight {
  component: string;
  configured_weight: number;
  enabled: boolean;
  effective_weight: number;
}

export interface ParsedCv {
  language_info: LanguageInfo;
  skills: string[];
  experience_summary: string | null;
  experience_evidence: string[];
  projects: string[];
  education: string[];
  certifications: string[];
  localized_evidence: LocalizedEvidence[];
}

export interface ParsedJd {
  language_info: LanguageInfo;
  required_skills: string[];
  preferred_skills: string[];
  unknown_skills: string[];
  skill_requirements: JdSkillRequirement[];
  job_title: string | null;
  responsibilities: string[];
  experience_requirements: string[];
  project_requirements: string[];
  education_requirements: string[];
  domain_keywords: string[];
}

export interface LanguageInfo {
  primary_language: string;
  detected_languages: string[];
  is_mixed_language: boolean;
  confidence: number;
}

export interface LocalizedEvidence {
  source_language: string;
  english_text: string | null;
}

export interface JdSkillRequirement {
  skill: string;
  classification: SkillPriority;
  critical: boolean;
  evidence: string | null;
  english_evidence: string | null;
  base_weight: number;
}

export interface SkillMatch {
  jd_skill: string;
  resume_skill: string | null;
  similarity: number;
  match_level: MatchLevel;
  importance: number;
  classification: SkillPriority;
  match_type: MatchType;
  evidence_score: number;
}

export interface SkillWeightSource {
  skill: string;
  jd_classification: SkillPriority;
  jd_base_weight: number;
  final_weight: number;
  cv_evidence_score: number;
  contribution: number;
}

export interface ScoreAdjustment {
  rule: string;
  reason: string;
  score_before: number;
  score_after: number;
}

export interface AnalysisResult {
  fit_score: number;
  fit_level: FitLevel;
  score_breakdown: ScoreBreakdown;
  parsed_cv: ParsedCv;
  parsed_jd: ParsedJd;
  skill_matches: SkillMatch[];
  matched_skills: string[];
  missing_skills: string[];
  cv_improvement_suggestions: string[];
  cover_letter: string | null;
  learning_roadmap: string[] | null;
  raw_fit_score: number;
  score_confidence: number;
  scoring_version: string;
  skill_weight_sources: SkillWeightSource[];
  score_adjustments: ScoreAdjustment[];
  analysis_warnings: string[];
}

export interface CompletedAnalysis {
  cv_parse_result: DocumentParseResult;
  jd_parse_result: DocumentParseResult | null;
  analysis_result: AnalysisResult;
}

export interface AnalysisSessionResponse {
  session_id: string;
  status: SessionStatus;
  result: CompletedAnalysis | null;
  error: SessionError | null;
  expires_at: string;
}
